All or Nothing Auth

NoopDog opened the pull request.

NoopDog committed:
wip

MillenniumFalconMechanic committed:
Upgraded Http to HttpClient. Added redirect to login on 403. Minor linting.

NoopDog commented:
@coverbeck this is not ready to merge but I wanted to let you know we are deep into getting this going. We updated the HTTPClient to the newer API along the way. We need to test as all of the DAOs have changed and we also need to remove the flash of the BW screen with no data before the auth wall shows. Picking this up agin in our morning and expect to be through it for your Friday am.

coverbeck commented:
@NoopDog Just curious, are you able to log in from your locally running UI using `https://jesse.ucsc-cgp-dev.org/` as the backend? I would think not, but maybe you've discovered a way, and I'm going down the wrong road in https://ucsc-cgl.atlassian.net/browse/TLP-559.

The first issue would be that the backend is only issuing cookies with the secure attribute set, so unless your local UI is on https I don't think it will work.

MillenniumFalconMechanic committed:
Fixed observable catch/throw imports.

MillenniumFalconMechanic committed:
Removed visual flash of file component content before auth has resolved.

MillenniumFalconMechanic committed:
Updated authenticated user header UI.

MillenniumFalconMechanic committed:
Added interceptor for 403.

MillenniumFalconMechanic committed:
Removed hard-coded test user name.

MillenniumFalconMechanic committed:
Disabled autocomplete on search type facet.

MillenniumFalconMechanic committed:
Added register component. Updated login component. Updated 401/403 to internal routes.

NoopDog committed:
text updates to login and register

NoopDog commented:
HI @coverbeck I have rebased on feature/commons and am submitting the PR. Can you push this to the "Charles" test instance and if it looks ok merge this into feature commons? 

Cheers and Thanks, D

coverbeck commented on spa/src/app/auth/error.interceptor.ts:
```diff
@@ -0,0 +1,47 @@
+                }
+                // Navigate to register on 403
+                else if ( response instanceof HttpErrorResponse && response.status === 403 ) {
+                    this.router.navigateByUrl("/register");
```
Could you redirect to `/unauthorized` instead? Or better yet, if it's not a hassle, `/unauthorized?account=<redacted email>`.

The problem is that we have the contact email configurable, via a property, and the currently configured value doesn't match your hard-coded email in register.component.html.

Our options are 

1. Change the hard-coded email in here to match the configured email, and update the source code every time it changes.
2. Surface the configured email via an API so you can display it dynamically
3. Just redirect to `/unauthorized`.

I think it's easiest to just do 3, and you can fix up that page in the Gatsby work, which will then apply to both being invoked from here and when an unauthorized user tries to log-in.

NoopDog replied:
Ok redirecting to /unauthorized for now.

coverbeck commented on spa/src/app/auth/register/register.component.html:
```diff
@@ -0,0 +1,8 @@
+        Register
+    </h1>
+    <p class="mat-body-1">
+        To register for data access, please contact <a href="mailto:casloan@ucsc.edu">casloan@ucsc.edu</a>.
```
See my above comment... I think it would best not to have this email merged into the main branch as well, as it might get harvested.

NoopDog replied:
will delete.

coverbeck commented on spa/src/app/data/user/user.service.ts:
```diff
@@ -1,33 +1,46 @@
      * @returns {Observable<void>}
      */
-    downloadRedwoodToken(): Observable<boolean> {
+    public downloadRedwoodToken(): Observable<boolean> {
```
This method can be removed if you want -- since Commons Boardwalk uses Blue Box and Azul, there is no longer Redwood.

Or you could just leave it for a future PR.

coverbeck reviewed (changes requested):
In addition to the 403 redirect comments, I think your copyright messages should say CGP, not CGL, and point to https://cgp.genomics.ucsc.edu, although it's not a big deal if you keep it.

CGL = Wet lab stuff
CGP = Computational stuff

coverbeck commented:
My comment about CGL being a wet lab was wrong; it is dry lab research. That said I still think you should CGP, as that is our team. Again, not a big deal.

NoopDog committed:
address pr comments

NoopDog commented:
Hi @coverbeck I believe I have addressed your PR comments. Please take a look. If If you can deploy to  the your server as well we can test this out and merge if it looks ok

coverbeck commented on spa/src/app/auth/error.interceptor.ts:
```diff
@@ -33,13 +35,15 @@ export class ErrorInterceptor implements HttpInterceptor {
                 // Navigate to register on 403
                 else if ( response instanceof HttpErrorResponse && response.status === 403 ) {
-                    this.router.navigateByUrl("/register");
+                    this.document.location.replace(`${rootUrl}/unathorized`);
```
Typo in `unauthorized`

NoopDog replied:
This should be addressed.. thanks!

NoopDog committed:
fix typo in error interceptor

NoopDog commented:
@coverbeck the URL typo should be addressed
